packages/xstate-solid/src/fsm.ts

```diff
@@ -0,0 +1,74 @@
+import type {
+  MachineImplementationsFrom,
+  ServiceFrom,
+  StateFrom,
+  StateMachine
+} from '@xstate/fsm';
+import { createMachine, interpret } from '@xstate/fsm';
+import type { Accessor } from 'solid-js';
+import { createMemo, onCleanup, createEffect } from 'solid-js';
+import { createImmutable } from './createImmutable';
+import { isServer } from 'solid-js/web';
+
+type UseFSMReturn<TService extends StateMachine.AnyService> = [
+  StateFrom<TService>,
+  TService['send'],
+  TService
+];
+
+export function useMachine<TMachine extends StateMachine.AnyMachine>(
+  machine: TMachine,
+  options?: MachineImplementationsFrom<TMachine>
+): UseFSMReturn<ServiceFrom<TMachine>> {
+  const resolvedMachine = createMachine(
+    machine.config,
+    options ? options : (machine as any)._options
+  );
+
+  const service = interpret(resolvedMachine);
+
+  if (!isServer) {
+    service.start();
+    onCleanup(() => service.stop());
+  }
+
+  return useService(service) as UseFSMReturn<ServiceFrom<TMachine>>;
+}
+
+export function useService<TService extends StateMachine.AnyService>(
+  service: TService | Accessor<TService>
+): UseFSMReturn<TService> {
+  const serviceMemo = createMemo(() =>
+    typeof service === 'function' ? service() : service
+  );
+
+  const initialService = serviceMemo();
+  const [state, setState] = createImmutable(
+    deriveFSMState(initialService.state as StateFrom<TService>)
+  );
+
+  createEffect(() => {
+    const currentService = serviceMemo();
+    const { unsubscribe } = currentService.subscribe(() =>
+      setState(deriveFSMState(currentService.state as StateFrom<TService>))
+    );
+    onCleanup(unsubscribe);
+  });
+
+  const send = (event: TService['send']) => serviceMemo().send(event);
+
+  return [state, send, serviceMemo()];
```

Comment:
I kinda think that the third element should not be returned from here. It could be achieved this way:
```diff
diff --git a/packages/xstate-solid/src/fsm.ts b/packages/xstate-solid/src/fsm.ts
index 5c3ac9faa..7b2710864 100644
--- a/packages/xstate-solid/src/fsm.ts
+++ b/packages/xstate-solid/src/fsm.ts
@@ -32,12 +32,13 @@ export function useMachine<TMachine extends StateMachine.AnyMachine>(
     onCleanup(() => service.stop());
   }
 
-  return useService(service) as UseFSMReturn<ServiceFrom<TMachine>>;
+  const [state, send] = useService(service);
+  return [state, send, service] as UseFSMReturn<ServiceFrom<TMachine>>;
 }
 
 export function useService<TService extends StateMachine.AnyService>(
   service: TService | Accessor<TService>
-): UseFSMReturn<TService> {
+): [StateFrom<TService>, TService['send']] {
   const serviceMemo = createMemo(() =>
     typeof service === 'function' ? service() : service
   );
@@ -55,7 +56,7 @@ export function useService<TService extends StateMachine.AnyService>(
 
   const send: TService['send'] = (event) => serviceMemo().send(event);
 
-  return [state, send, serviceMemo()];
+  return [state, send];
 }
 
 function deriveFSMState<State extends StateMachine.AnyState>(
```

OTOH - it is actually returned in `@xstate/react/fsm` so perhaps we should leave this as it here and fix this together in the future
https://github.com/statelyai/xstate/blob/afc7a83f9cd33af6b5480359f78f786d217bb0fd/packages/xstate-react/src/fsm.ts#L122

Reply:
OTOH - if we return it here, shouldn't it be a lazy function and not a resolved one?
```diff
-  return [state, send, serviceMemo()];
-  return [state, send, serviceMemo];
```

Reply:
Either way works fine; I resolve the function, so the return type is the same as core useMachine. It makes it easier for downstream typing with the service and not within a lazy function. I don't have a strong opinion on this, though, so happy to switch it if you think it's worthwhile.

Reply:
> Either way works fine; I resolve the function, so the return type is the same as core useMachine.

That makes sense - as in, it would be nice if the return types would match. However, `useMachine` always returns the same service and `useService` can switch between different services - so, if I understand this correctly, if we return `serviceMemo()` from the `useService` then it might become "stale", right?

Perhaps the easiest solution here is to actually drop this return value from the `useService` after all. It's not needed there since it's already its argument - we don't have to return to the user what they already have.